client/web/src/components/popovers/ColorPicker.vue

```diff
@@ -115,13 +117,197 @@
+				this.setHuePosition(e.y - colorPicker.hue.rect.top);
+			} else if (this.state === "move_opacity") {
+				this.setOpacityPosition(e.y - colorPicker.opacity.rect.top);
+			} else if (this.state === "move_saturation") {
+				this.setSaturationPosition(e.x - colorPicker.saturation.rect.left, e.y - colorPicker.saturation.rect.top);
+			}
+
+			if (this.state !== "idle") {
+				this.updateHue();
+				this.$emit("update:color", HSV2RGB(colorPicker.color));
+			}
+		},
+
+		onPointerUp() {
+			if (this.state !== "idle") {
+				this.state = "idle";
+				this.removeEvents();
+			}
+		},
+
+		updateRects() {
+			const { colorPicker } = this.$options as { colorPicker: ColorPickerData };
+
+			const saturationPicker = this.$refs.saturationPicker as HTMLDivElement;
+			const saturation = saturationPicker.getBoundingClientRect();
+			colorPicker.saturation.rect.width = saturation.width;
+			colorPicker.saturation.rect.height = saturation.height;
+			colorPicker.saturation.rect.left = saturation.left;
+			colorPicker.saturation.rect.top = saturation.top;
+
+			const huePicker = this.$refs.huePicker as HTMLDivElement;
+			const hue = huePicker.getBoundingClientRect();
+			colorPicker.hue.rect.width = hue.width;
+			colorPicker.hue.rect.height = hue.height;
+			colorPicker.hue.rect.left = hue.left;
+			colorPicker.hue.rect.top = hue.top;
+
+			const opacityPicker = this.$refs.opacityPicker as HTMLDivElement;
+			const opacity = opacityPicker.getBoundingClientRect();
+			colorPicker.opacity.rect.width = opacity.width;
+			colorPicker.opacity.rect.height = opacity.height;
+			colorPicker.opacity.rect.left = opacity.left;
+			colorPicker.opacity.rect.top = opacity.top;
+		},
+
+		setSaturationPosition(x: number, y: number) {
+			const { colorPicker } = this.$options as { colorPicker: ColorPickerData };
+			const saturationCursor = this.$refs.saturationCursor as HTMLDivElement;
+			const saturationPosition = [clamp(x, 0, colorPicker.saturation.rect.width), clamp(y, 0, colorPicker.saturation.rect.height)];
+			saturationCursor.style.transform = `matrix(1, 0, 0, 1, ${saturationPosition[0]}, ${saturationPosition[1]})`;
+			colorPicker.color.s = saturationPosition[0] / colorPicker.saturation.rect.width;
+			colorPicker.color.v = (1 - saturationPosition[1] / colorPicker.saturation.rect.height) * 255;
+		},
+
+		setHuePosition(y: number) {
+			const { colorPicker } = this.$options as { colorPicker: ColorPickerData };
+			const hueCursor = this.$refs.hueCursor as HTMLDivElement;
+			const huePosition = clamp(y, 0, colorPicker.hue.rect.height);
+			hueCursor.style.transform = `matrix(1, 0, 0, 1, 0, ${huePosition})`;
+			colorPicker.color.h = clamp(1 - huePosition / colorPicker.hue.rect.height);
+		},
+
+		setOpacityPosition(y: number) {
+			const { colorPicker } = this.$options as { colorPicker: ColorPickerData };
+			const opacityCursor = this.$refs.opacityCursor as HTMLDivElement;
+			const opacityPosition = clamp(y, 0, colorPicker.opacity.rect.height);
```

Comment:
It might be cleaner to use percentages instead of pixels. But that's not a hard requirement, just a suggestion.